### PR TITLE
don't fetch indices , add indices to cache, and add esmonthly func

### PR DESCRIPTION
**Pull request by kylebrandt**

### CODE REVIEW

**Comment by kylebrandt** [comment]
Currently if the time series is missing it errors inside ESDateHistrogram when it can't get the expected time series in the result. This result gets cached. Before when it would error on no indices in `getFn`, a wrapper for  `func (e ElasticHosts) Query(r *ElasticRequest) (*elastic.SearchResult, error)`, ensuring it was not cached. But now a lack of indices throws an error. Need to research inspecting the shards of the result and understanding what the means in this context.

**Comment by kylebrandt** [comment]
I think counting the resulting shards and erroring when there were no shards that matched will get the result I want. 

However, it depends on the following PR https://github.com/olivere/elastic/pull/373 

Alternatively, I could go for the date math method mentioned in https://github.com/bosun-monitor/bosun/issues/1926#issuecomment-252287736 . But only if `allow_no_indices` applies to datemath and not just wildcard. According to the documentation it only applies to wildcard which I haven't tested yet. Even if does work I think the implementation will be a bit of pain.

**Comment by kylebrandt** [comment]
note: forgot to fmt, force pushed the fmt.

File: cmd/bosun/expr/elastic.go [hunk 1]
```diff
@@ -265,37 +269,47 @@ func (e ElasticHosts) Query(r *ElasticRequest) (*elastic.SearchResult, error) {
...
-	var ok bool
-	str, ok = s.(string)
-	if !ok {
-		return "", fmt.Errorf("failed to generate string representation of search source for cache key: %s", s)
+	b, err := json.Marshal(s)
+	if err != nil {
+		return "", fmt.Errorf("failed to generate json representation of search source for cache key: %s", s)
 	}
-	return str, nil
+	return fmt.Sprintf("%v\n%s", r.Indices, b), nil
```
Comment (captncraig) [inline cmd/bosun/expr/elastic.go]: probably use `string(b)` here. Don't want `[42 31 13]` in you key.

Comment (captncraig) [inline cmd/bosun/expr/elastic.go]: additionally, what does the source have that causes it to be a different key? What varies?

Comment (captncraig) [inline cmd/bosun/expr/elastic.go]: I ask because it surprises me that something like getting a cache key even has a possibility of erroring.

**Comment by kylebrandt** [comment]
fixes #1926

**Pull request merged**